server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/MasterHistoryAction.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.admin.cluster.coordination;
+
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.support.master.MasterNodeReadRequest;
+import org.elasticsearch.cluster.coordination.MasterHistory;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * This action is used to fetch the MasterHistory from a remote node.
+ */
+public class MasterHistoryAction extends ActionType<MasterHistoryAction.Response> {
+
+    public static final MasterHistoryAction INSTANCE = new MasterHistoryAction();
+    public static final String NAME = "cluster:internal/master_history/get";
+
+    private MasterHistoryAction() {
+        super(NAME, MasterHistoryAction.Response::new);
+    }
+
+    public static class Request extends MasterNodeReadRequest<MasterHistoryAction.Request> {
```

Comment:
I believe this could be a `ActionRequest` and receive a `DiscoveryNode` as a parameter (just to be sure we're getting the history from the node we want - if there's a master change in the meantime we might be getting the history from another node by the time we submit this) https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/cache/FrozenCacheInfoNodeAction.java is a good example

cc @DaveCTurner

Reply:
More generally I'd suggest adding the code that uses this machinery to this same PR. I think there will be other opportunities to simplify things and perhaps change the message flow but it's hard to see what's needed without the calling code too.

(Acking that this'll make the PR bigger, and we might want to split it up again later, but for now I'd rather see everything)

Reply:
> (Acking that this'll make the PR bigger, and we might want to split it up again later, but for now I'd rather see everything)

Ha yeah originally the calling code was in this PR and I pulled it out. I'll put it back in for now.

Reply:
> I believe this could be a `ActionRequest`

Oops I had originally incorrectly had this as a TransportMasterNodeReadAction and forgot to fix the Request when I fixed the rest. Fixing that now.

> and receive a `DiscoveryNode` as a parameter

I don't think I'm following you here. You mean as an argument to its constructor rather than calling `remoteAddress()` directly? Or something else?